pkg/maps/nat/stats/stats.go

```diff
@@ -0,0 +1,291 @@
+		maxPorts: maxAvailPorts,
+		db:       params.DB,
+		table:    params.Table,
+	}
+	statsCtx, cancelStats := context.WithCancel(context.Background())
+	params.Lifecycle.Append(cell.Hook{
+		OnStart: func(hc cell.HookContext) error {
+			ctx, cancel := context.WithTimeout(context.Background(), time.Second*120)
+			defer cancel()
+			nmap4, err := params.NatMap4.Await(ctx)
+			if err != nil {
+				if !errors.Is(err, nat.MapDisabled) {
+					return err
+				}
+			}
+			nmap6, err := params.NatMap6.Await(ctx)
+			if err != nil {
+				if !errors.Is(err, nat.MapDisabled) {
+					return err
+				}
+			}
+			m.natMap4 = nmap4
+			m.natMap6 = nmap6
+			if m.natMap4 == nil && m.natMap6 == nil {
+				return nil
+			}
+
+			m.run(statsCtx, params.Health.NewScope("nat-stats"))
+			return nil
+		},
+		OnStop: func(hc cell.HookContext) error {
+			cancelStats()
+			return nil
+		},
+	})
+	return m, nil
+}
+
+func upsertStat[KT tupleKey](m *Stats, topk *topk[KT], family nat.IPFamily) error {
+	tx := m.db.WriteTxn(m.table)
+	defer tx.Abort()
+
+	var errs error
+	iter, _ := m.table.All(tx)
+	for entry, _, ok := iter.Next(); ok; entry, _, ok = iter.Next() {
+		if entry.Type == family.String() {
+			_, _, err := m.table.Delete(tx, entry)
+			errors.Join(errs, err)
+		}
+	}
+
+	topk.popForEach(func(key KT, count, ith int) {
+		if ith == kBucketsToStore-1 {
+			m.metrics.updateLocalPorts(family, count, m.maxPorts)
+		}
+		extip, eip, proto, rport := key.tuple()
+		_, _, err := m.table.Insert(tx, NatMapStats{
+			Type:       family.String(),
+			EgressIP:   eip,
+			EndpointIP: extip,
+			RemotePort: rport,
+			Proto:      proto,
+			Count:      count,
+			Nth:        ith,
+		})
+		if err != nil {
+			errs = errors.Join(errs, err)
+		}
+	})
+	if errs != nil {
+		return fmt.Errorf("failures occured updating stats table, transaction will not be committed: %w", errs)
+	}
+	tx.Commit()
+	return nil
+}
+
+func (m *Stats) run(ctx context.Context, hr cell.Health) {
+	controller.NewManager().UpdateController("nat-stats", controller.ControllerParams{
+		RunInterval: m.config.NATMapStatInterval,
+		Context:     ctx,
+		Health:      hr,
+		Group:       controller.NewGroup("nat-stats"),
+		DoFunc:      m.countNat,
+	})
+}
+
+func swapBytes(n uint16) uint16 {
+	return (n<<8)&0xff00 | (n>>8)&0x00ff
+}
+
+func (m *Stats) countNat(ctx context.Context) error {
+	var errs error
+	if m.natMap4 != nil {
+		tupleToPortCount := make(map[tuple.TupleKey4]uint16, 128)
+		_, err := m.natMap4.ApplyBatch4(func(keys []tuple.TupleKey4, vals []nat.NatEntry4, size int) {
+			for i := 0; i < size; i++ {
+				key := keys[i]
+				key.SourcePort = swapBytes(key.SourcePort)
+				if key.Flags == tuple.TUPLE_F_IN &&
+					(key.NextHeader == u8proto.TCP || key.NextHeader == u8proto.ICMP) {
+					key.DestPort = 0
+					ports := tupleToPortCount[key]
+					ports++
+					tupleToPortCount[key] = ports
+				}
+			}
+		})
+		if err != nil {
+			m.logger.WithError(err).Error("failed to count ipv4 nat map entries")
```

Comment:
This will be a user-facing error log msg, so I think it should provide a bit more context as to the consequences of erroring here, and what might the user want to do or what Cilium will do about it.

Reply:
Same applies to other log msgs with WARN level and above.

Reply:
good point